src/main/java/graphql/execution/reactive/ReactiveSupport.java

```diff
@@ -0,0 +1,176 @@
+package graphql.execution.reactive;
+
+import graphql.DuckTyped;
+import graphql.Internal;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Flow;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This provides support for a DataFetcher to be able to
+ * return a reactive streams {@link Publisher} or Java JDK {@link Flow.Publisher}
+ * as a value, and it can be turned into a {@link CompletableFuture}
+ * that we can get an async value from.
+ */
+@Internal
+public class ReactiveSupport {
+
+    @DuckTyped(shape = "CompletableFuture | Object")
+    public static Object fetchedObject(Object fetchedObject) {
+        if (fetchedObject instanceof Flow.Publisher) {
+            return flowPublisherToCF((Flow.Publisher<?>) fetchedObject);
+        }
+        if (fetchedObject instanceof Publisher) {
+            return reactivePublisherToCF((Publisher<?>) fetchedObject);
+        }
+        return fetchedObject;
+    }
+
+    private static CompletableFuture<Object> reactivePublisherToCF(Publisher<?> publisher) {
+        ReactivePublisherToCompletableFuture<Object> cf = new ReactivePublisherToCompletableFuture<>();
+        publisher.subscribe(cf);
+        return cf;
+    }
+
+    private static CompletableFuture<Object> flowPublisherToCF(Flow.Publisher<?> publisher) {
+        FlowPublisherToCompletableFuture<Object> cf = new FlowPublisherToCompletableFuture<>();
+        publisher.subscribe(cf);
+        return cf;
+    }
+
+    /**
+     * The implementations between reactive Publishers and Flow.Publishers are almost exactly the same except the
+     * subscription class is different.  So this is a common class that contains most of the common logic
+     *
+     * @param <T> for two
+     * @param <S> for subscription
+     */
+    private static abstract class PublisherToCompletableFuture<T, S> extends CompletableFuture<T> {
```

Comment:
This class is heavily inspired from the Reactor `reactor.core.publisher.MonoToCompletableFuture` class.  Its uses pretty much the same pattern to turn a Publisher into a CF

Reply:
Note it "IS-A" CompletableFuture class